File: RecallThatAndroid/app/src/main/java/com/recallthat/Services/ocr/OCRPipelineService.kt
package com.recallthat.Services.ocr

import android.net.Uri
import com.recallthat.Models.OCRStatus
import com.recallthat.Persistence.MemoryRepository

data class OCRProgress(val completed: Int, val total: Int)

class OCRPipelineService(
    private val ocrService: OCRService,
    private val repository: MemoryRepository
) {
    suspend fun processQueue(onProgress: (OCRProgress) -> Unit) {
        val pending = repository.fetchAll().filter { it.ocrStatus == OCRStatus.NOT_STARTED }
        val total = pending.size
        var completed = 0

        for (memory in pending) {
            val uri = memory.photoAssetUri ?: continue
            val withPending = memory.copy(ocrStatus = OCRStatus.PENDING)
            repository.update(withPending)

            val (newStatus, text) = try {
                val extracted = ocrService.extractText(Uri.parse(uri))
                OCRStatus.COMPLETE to extracted
            } catch (e: Exception) {
                OCRStatus.FAILED to ""
            }

            val title = text.lines()
                .firstOrNull { it.trim().length >= 3 }
                ?.trim()
                ?: ""
            val searchText = "$title $text".lowercase()

            repository.update(
                memory.copy(
                    ocrStatus = newStatus,
                    ocrText = text,
                    title = title,
                    searchText = searchText
                )
            )
            completed++
            onProgress(OCRProgress(completed, total))
        }
    }
}

File: RecallThat/Services/OCR/OCRPipelineService.swift
import Foundation
import UIKit

/// Processes the queue of memories waiting for OCR.
/// Works through notStarted memories one at a time, updating status and text in the repository.
@MainActor
final class OCRPipelineService {
    private let ocrService: any OCRServiceProtocol
    private let repository: any MemoryRepository

    init(ocrService: any OCRServiceProtocol, repository: any MemoryRepository) {
        self.ocrService = ocrService
        self.repository = repository
    }

    /// Processes all memories with ocrStatus == .notStarted that have a photo asset.
    /// Calls onStart with the total count, then onProgress after each item.
    func processQueue(onStart: ((Int) -> Void)? = nil, onProgress: ((UUID) -> Void)? = nil) async {
        guard let memories = try? await repository.fetchAll() else { return }
        let queue = memories.filter {
            $0.ocrStatus == .notStarted && $0.photoAssetIdentifier != nil
        }
        guard !queue.isEmpty else { return }

        // Keep running for ~30 s after the app backgrounds so the queue doesn't stall mid-batch.
        var bgTask = UIBackgroundTaskIdentifier.invalid
        bgTask = UIApplication.shared.beginBackgroundTask(withName: "RecallThat.OCR") {
            UIApplication.shared.endBackgroundTask(bgTask)
        }
        defer { UIApplication.shared.endBackgroundTask(bgTask) }

        onStart?(queue.count)
        for memory in queue {
            await processOne(memory: memory)
            onProgress?(memory.id)
        }
    }

    // MARK: - Private

    private func processOne(memory: Memory) async {
        guard let identifier = memory.photoAssetIdentifier else { return }

        var updated = memory
        updated.ocrStatus = .pending
        try? await repository.update(updated)

        do {
            let text = try await ocrService.extractText(from: identifier)
            updated.ocrText = text
            updated.ocrStatus = .complete
            if updated.title.isEmpty {
                updated.title = Self.firstMeaningfulLine(from: text)
            }
            updated.searchText = Self.buildSearchText(title: updated.title, ocrText: text)
        } catch {
            updated.ocrStatus = .failed
        }

        try? await repository.update(updated)
    }

    // MARK: - Helpers

    static func firstMeaningfulLine(from text: String) -> String {
        let line = text
            .components(separatedBy: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .first { $0.count >= 3 } ?? ""
        return String(line.prefix(60))
    }

    static func buildSearchText(title: String, ocrText: String) -> String {
        "\(title) \(ocrText)".lowercased()
    }
}
